Polio-1253: Refactor campaigns front-end

quang-le opened the pull request:
Related JIRA tickets :POLIO-1253

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- convert `Dashboard` to tsx
- remove modals from `Dashboard`'s jsx
- remove useless state from Dashboard.tsx
- move table columns code to separate hook
- convert `CreateEditDialog` to tsx, move tabs code into a hook for legibility
- rename `PageActionWithLink` as `LinkButton`
- remove `ConfirmDialog`, `PageActions` and `PageAction` components
- replace `PageActions` with `DashboardButtons`
- move edit, delete and restore dialogs in table columns hook
- change parameters of `useGetCampaigns` for smoother user experience
- improve loading state of the Create and Edit modals to show spinner while saving
- improve loading state of the table to show spinner when deleting/restoring a campaign
- align csv button with table and filter button

## How to test

Go to campaigns page, try functionalities. Nothing should have changed
Go to grouped campaigns and try outthe link. You should always have the "edit" modal open

## Print screen / video

![Screenshot 2023-12-21 at 15 41 20](https://github.com/BLSQ/iaso/assets/38907762/f3aac089-9592-44e5-aa5f-6b53fa6eaf97)

## Notes

The initial ticket was about links to campaigns that would sometimes open the create modal i.o the edit modal of the selected campaign. The refactor relies only on url params to open the edit modal (no internal state), so it should be fixed. It's also hard to test since the initial bug is kinda flaky

Needs to be merged after https://github.com/BLSQ/iaso/pull/989

DO NOT MERGE UNTIL CYPRESS TESTS ARE WRITTEN FOR THIS PAGE

quang-le committed:
IA-2570: add delete and restore modals

quang-le committed:
POLIO-1253: move tab state out of Dashboard

quang-le committed:
POLIO-1253: convert config to tsx

quang-le committed:
POLIO-1253: convert CreateEditDialog to tsx

quang-le committed:
POLIO-1253: convert Dashboard to tsx

- move buttons (create, export) to own component
- move table columns to own hook
- rename PageActionWithLink as LinkButton
- use DeleteModal and RestoreModal, and remove related code from state

quang-le committed:
POLIO-1253: remove unused css classes and components

quang-le committed:
IA-2570: make children optional

quang-le committed:
Merge branch 'IA-2570_delet_restore_modals' into POLIO-1253_broken_links_to_campaign

quang-le committed:
POLIO-1357: refactor CampaignModal WIP

quang-le committed:
Merge remote-tracking branch 'origin/main' into POLIO-1253_broken_links_to_campaign

quang-le committed:
POLIO-1253: update mui imports

quang-le committed:
POLIO-1253: use EditMocal in campaigns table

quang-le committed:
POLIO-1253: code cleanup

quang-le committed:
POLIO-1253: improve EditModal open/close speed

quang-le committed:
POLIO-1253: use CreateCampaignModal in Dashboard

quang-le committed:
POLIO-1253: cleanup comments

quang-le committed:
POLIO-1253: add staleTime to useGetCampaigns

quang-le committed:
POLIO-1253: fix dashboard buttons positioning

quang-le committed:
POLIO-1253: add spacing between buttons and table

quang-le commented on plugins/polio/js/src/components/Buttons/LinkButton.tsx:
```diff
@@ -8,20 +8,15 @@ type Props = {
-            <Button
-                variant="contained"
-                color="primary"
-                className={classes.pageAction}
```
This class didn't exist

quang-le commented on plugins/polio/js/src/domains/Campaigns/MainDialog/CreateEditDialog.tsx:
```diff
@@ -254,24 +200,6 @@ const CreateEditDialog = ({ isOpen, onClose, campaignId }) => {
                     <Form>
                         <CurrentForm />
                     </Form>
-                    <Grid container justifyContent="flex-end">
-                        <Grid item md={6}>
-                            {formik.errors?.non_field_errors?.map(
-                                (error_msg, i) => (
-                                    <Typography key={i} color="error">
-                                        {error_msg}
-                                    </Typography>
-                                ),
-                            )}
-                        </Grid>
-                    </Grid>
```
this was mapping non_field errors, but formik expects it to be a string

quang-le commented on plugins/polio/js/src/domains/Campaigns/Rounds/RoundsEmptyDates.tsx:
dead code

quang-le commented on plugins/polio/js/src/domains/Campaigns/hooks/api/useGetCampaigns.ts:
```diff
@@ -127,6 +127,7 @@ export const useGetCampaigns = (
         queryFn: () => getRequest(getURL(params, url)),
         options: {
             cacheTime: Infinity,
+            staleTime: 1000 * 60 * 15,
```
I added this to be in line with other use-query hooks

quang-le commented on plugins/polio/js/src/domains/Campaigns/MainDialog/CreateEditDialog.tsx:
```diff
@@ -61,13 +63,18 @@ const CreateEditDialog = ({ isOpen, onClose, campaignId }) => {
             onSuccess: () => {
                 helpers.resetForm();
                 onClose();
+                dispatch(
+                    enqueueSnackbar(
+                        successSnackBar(MESSAGES.defaultMutationApiSuccess),
+                    ),
+                );
```
dispatching the snackbar here so the form can be reset and `onClose` called before the snacbar triggers a re-render of the whole table

quang-le committed:
POLIO-1253: add explanatory comments

quang-le commented:
Converted back to draft because calendar code should also be updated

quang-le committed:
Merge branch 'main' into POLIO-1253_broken_links_to_campaign

beygorghor commented on plugins/polio/js/config.tsx:
```diff
@@ -19,11 +19,11 @@ import WatchLaterIcon from '@mui/icons-material/WatchLater';
 import StorageIcon from '@mui/icons-material/Storage';
 import NotificationsActiveIcon from '@mui/icons-material/NotificationsActive';
-import { Dashboard } from './src/domains/Campaigns/Dashboard';
+import { Dashboard } from './src/domains/Campaigns/Dashboard/Dashboard';
```
not a fan ...

quang-le replied:
And I don't like `index` files 😅. I can try and look for a better name

quang-le replied:
renamed the folder

beygorghor commented on plugins/polio/js/config.tsx:
```diff
@@ -48,30 +48,32 @@ import {
     STOCK_VARIATION,
     NOTIFICATIONS_BASE_URL,
 } from './src/constants/routes';
+// @ts-ignore
```
we can add a config in tsConfig to include json files:
```
{
  "compilerOptions": {
    "resolveJsonModule": true,
      ... 
  }
  ...
}
```

beygorghor commented on plugins/polio/js/src/domains/Campaigns/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,112 @@
+import { Filters } from '../../Calendar/campaignCalendar/Filters';
+import { DashboardButtons } from './DashboardButtons';
+
+const Dashboard = ({ router }) => {
```
No typing here ?

beygorghor commented on plugins/polio/js/src/domains/Campaigns/Dashboard/DashboardButtons.tsx:
```diff
@@ -0,0 +1,47 @@
+
+const useStyles = makeStyles(theme => {
+    return {
+        marginBottom: { marginBottom: theme.spacing(0) },
```
better use a Box with mb prop

quang-le replied:
Question which is the most costly: adding a css rule, or adding a level of nesting in the dom with the Box?

I don't really mind either way. I thought I'd try this way to avoid adding a level of nesting just for a margin bottom

beygorghor commented on plugins/polio/js/src/domains/Campaigns/Dashboard/useCampaignsTableColumns.tsx:
```diff
@@ -0,0 +1,132 @@
+    // eslint-disable-next-line no-unused-vars
+    handleClickRestoreRow: (value: any) => void;
+    // eslint-disable-next-line no-unused-vars
+    handleClickDeleteRow: (value: any) => void;
```
value is an ID, a number i think

beygorghor commented on plugins/polio/js/src/domains/Campaigns/Dashboard/useCampaignsTableColumns.tsx:
```diff
@@ -0,0 +1,132 @@
+                            </span>
+                        </Tooltip>
+                    );
+                },
```
this looks weird, did you put the code of cell in accessor ?

quang-le replied:
Honestly I just pasted the existing code in a hook

beygorghor commented on plugins/polio/js/src/domains/Campaigns/MainDialog/CreateEditDialog.tsx:
```diff
@@ -115,69 +125,7 @@ const CreateEditDialog = ({ isOpen, onClose, campaignId }) => {
-        formik.values.rounds.length,
-        selectedCampaign?.rounds,
-    ]);
+    const tabs = usePolioDialogTabs(formik, selectedCampaign);
```
❤️

beygorghor commented on plugins/polio/js/src/domains/Campaigns/MainDialog/CreateEditDialog.tsx:
```diff
@@ -283,7 +213,7 @@ const CreateEditDialog = ({ isOpen, onClose, campaignId }) => {
                 </Button>
                 <Button
-                    onClick={formik.handleSubmit}
+                    onClick={() => formik.handleSubmit()}
```
why ?

quang-le replied:
There's a weird bug otherwise. Don't remember what exactly, but it's necessary

beygorghor reviewed (changes requested):
Can you give more padding to the dialog please ?
<img width="329" alt="Screenshot 2024-01-10 at 09 48 59" src="https://github.com/BLSQ/iaso/assets/12494624/708bda65-2b23-446c-963a-bb560e738a18">
<img width="1476" alt="Screenshot 2024-01-10 at 09 48 36" src="https://github.com/BLSQ/iaso/assets/12494624/a25de4f6-5385-4886-8705-50d9b4443ed2">
History button was smaller before

beygorghor commented:
Very nice job btw !

quang-le committed:
Merge remote-tracking branch 'origin/main' into python_39_POLIO-1253

quang-le committed:
POLIO-1253: remove ts error on json import

quang-le committed:
POLIO-1253: rename Dashboard folder, fix table cols

- remove cell code from accessor on columns hook
- rename Dashboard folder: CampaignsList

quang-le committed:
POLIO-1253: add padding to dialog

quang-le committed:
POLIO-1253: fix EditModal bugs

- fix backdrop click
- fix cancel button
- fix open issue with icon button in table
- clean up comment on import

beygorghor reviewed (approved):
we can make a separate ticket for that but:

- <img width="800" alt="Screenshot 2024-01-10 at 11 42 22" src="https://github.com/BLSQ/iaso/assets/12494624/6eab7ed7-bd39-49b9-a820-045ed7059af9">
- Space between title and tabs is too big
<img width="1441" alt="Screenshot 2024-01-10 at 11 42 32" src="https://github.com/BLSQ/iaso/assets/12494624/ce5fb3f6-fff0-4197-99b9-619405b1a75d">
- Space between text and close button is too small 
<img width="168" alt="Screenshot 2024-01-10 at 11 42 57" src="https://github.com/BLSQ/iaso/assets/12494624/6d6a4948-064c-419c-95a5-18fc21dc227b">

quang-le commented:
> we can make a separate ticket for that but:
> 
>     * <img alt="Screenshot 2024-01-10 at 11 42 22" width="800" src="https://private-user-images.githubusercontent.com/12494624/295538364-6eab7ed7-bd39-49b9-a820-045ed7059af9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4ODcxMjksIm5iZiI6MTcwNDg4NjgyOSwicGF0aCI6Ii8xMjQ5NDYyNC8yOTU1MzgzNjQtNmVhYjdlZDctYmQzOS00OWI5LWE4MjAtMDQ1ZWQ3MDU5YWY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDExNDAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1MDU2MTQxZWFmYmU5N2U4ZTBmYTBkMjZhMzg2ZWMzOGI0MTQ4MDBhZTdkNjM0MGUwYWZmNzIzZjkwMTU5YTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KrpOzsqXmogk8qGembM48ZelJ1MApoIr-Szjsn0GDRs">
> 
>     * Space between title and tabs is too big
> 
> 
> <img alt="Screenshot 2024-01-10 at 11 42 32" width="1441" src="https://private-user-images.githubusercontent.com/12494624/295538423-ce5fb3f6-fff0-4197-99b9-619405b1a75d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4ODcxMjksIm5iZiI6MTcwNDg4NjgyOSwicGF0aCI6Ii8xMjQ5NDYyNC8yOTU1Mzg0MjMtY2U1ZmIzZjYtZmZmMC00MTk3LTk5YjktNjE5NDA1YjFhNzVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDExNDAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViOTkzZWU0MzIwMGQ2NTExNzNiYjE4NTQwZmJlODFmMjZjZjZkM2NmM2MwYTU5MDdhMzlhOWViN2FmMzhiMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eJc5WPcL4G6598sMeD_B9zSyOy27aky_DDuLwpjzygQ"> - Space between text and close button is too small <img alt="Screenshot 2024-01-10 at 11 42 57" width="168" src="https://private-user-images.githubusercontent.com/12494624/295538683-6d6a4948-064c-419c-95a5-18fc21dc227b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4ODcxMjksIm5iZiI6MTcwNDg4NjgyOSwicGF0aCI6Ii8xMjQ5NDYyNC8yOTU1Mzg2ODMtNmQ2YTQ5NDgtMDY0Yy00MTljLTk1YTUtMThmYzIxZGMyMjdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDExNDAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4MTVlNWUzMTY4ZDI2NTVhY2Y0OTE2ZmUwMzUzZGNiNWIyNzY5NWU5YTUxZGNjZWMyNGFlNGJkNzY5ZGVlNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iGEOAV9Yfap4C7rDTV0e150q5xaFRpMnC-JtxaED-fg">

That looks more like an MUI5 issue to me. I didn't touch these components in the refactor

Ticket: https://bluesquare.atlassian.net/browse/POLIO-1404

quang-le committed:
Merge remote-tracking branch 'origin/main' into POLIO-1253_broken_links_to_campaign

quang-le committed:
POLIO-1253: fix conflict wit main
